homework_1/02_ProgressiveTax/taxes/taxes.go

```diff
@@ -0,0 +1,75 @@
+package taxes
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+type bracket struct {
+	threshold float32
+	taxRate   float32
+}
+
+func NewBracket(threshold, taxRate float32) (*bracket, error) {
```

Comment:
I believe this one can be private, I don't see it being useful to a user at all.

Reply:
Yeah, I had that in mind, but even though we don't use it in this simple task, generally, it would serve as a constructor for a single bracket and also take care of bracket validation. It is basically a connection between user and bracket struct, because bracket is not made public. That is because there has to be some sort of validation before creating a bracket, as it makes no sense when values are for example negative.

Reply:
But there's no public function that receives `*bracket` as an argument at the moment so I can't use the returned value in any way. Only thing I can do with is validate `threshold` and `taxRate` which isn't really useful and its name is `NewBracket` not `ValidateBracket`.

Reply:
Of course, but if we were to upgrade our system, that is the function that would have to be made public at some point. Anyway, I changed it to be private.

Reply:
> that is the function that would have to be made public at some point.

You could literally say that for almost any private function/method/class in any library. Maintaining a public name in a library is much harder than a private one so always strive to expose the minimal amount of names in your libraries.

Let's say you keep this `NewBracket` public and release your library to the world. The method is essentially useless to the users and most won't even use it but there's always someone, somewhere that will use it in the weirdest way possible. In one moment you need to update your library with a new feature and that requires you to change the prototype of the `NewBracket` function but you can't do that now because it's public - if you change the prototype you've broken the code of those couple of users that were using it. If you've kept it private you can freely change it (and then possibly expose a useful version of it as public).

Reply:
Alright, I understand. :smile: